### DISCUSSION TITLE
Windows and WASM performance

### DISCUSSION THREAD

**Discussion by benc-uk** [hajimehoshi/ebiten#1677] [Q&A] [answered]
I'm building a CHIP-8 emulator in Go using ebiten

I've spent quite a lot of time optimizing performance, esp for my rendering and how I output the display. My graphics are _incredibly_ basic, just a monochrome 64x32 image (1-bit so pixels are either on or off) which is then output scaled to the screen inside the Draw() function. I'm seeing a few things

I'm seeing a couple of things:
 - On Linux (WSL2) everything performs extremely well, so much so I have to slow the emulator processor loop down quite a lot to make programs playable
 - Windows is **significantly** slower, I have to set vsync disabled and ebiten.UncappedTPS set to make things playable even then it's not great. The difference here is not slight, it's probably several orders of magnitude, which I don't think is down to my code.
 - WASM is even slower still, I'm using 1000 x 500 output size which doesn't seem too outrageous, but nothing is really playable, everything just crawls 

My question is ebiten really only optimised for Linux? Should I continue trying to tune my code or is the slower performance on Windows and WASM a known problem?

**Comment by hajimehoshi** [comment, marked as answer]
> My question is ebiten really only optimised for Linux?

No, the program should work well on Windows as well. Could you share the code? Thanks!

**Reply by benc-uk** [reply]
Project is here https://github.com/benc-uk/chip8 the ebiten stuff is found in pkg/emulator

Summary of my code is

Emulator loop is started in a separate goroutine (I don't know if this is part of the problem), and Update() is used to check a channel to exit when the main emulator loop sends an error. Mostly I do nothing inside Update()

My `chip8Emulator` struct implements the `Game` interface and it holds it's own ebiten image in a field called `display`, it's just 64x32 pixels. Only two instructions in CHIP8 modify the display (CLR and DRW) so I only update the display when they have been called. The vm in the snippet below is the CHIP-8 virtual machine, it holds an internal 63x32 array for the display, the values of which are either 0 or 1

```go
// This is my implementation of the ebiten.Game interface
type chip8Emulator struct {
	vm        *chip8.VM // This is the CHIP-8 virtual machine
	errorChan chan error
	display   *ebiten.Image // internal display image
	pixelSize int
}
```
My Draw function is very simple
```go
func (e *chip8Emulator) Draw(screen *ebiten.Image) {
	if e.vm.DisplayUpdated {
		e.display.Clear()
		for y := 0; y < 32; y++ {
			for x := 0; x < 64; x++ {
				if e.vm.DisplayValueAt(x, y) == 1 {
					e.display.Set(x, y, pixelColour)
				}
			}
		}
		e.vm.DisplayUpdated = false
	}

        // Draw the internal emulator display into the screen and scale it up
	opt := &ebiten.DrawImageOptions{}
	opt.GeoM.Scale(float64(e.pixelSize), float64(e.pixelSize))
	opt.Filter = ebiten.FilterNearest
	screen.DrawImage(e.display, opt)
}
```

**Reply by benc-uk** [reply]
Oh and I'm building everything in Linux (WSL2) including the Windows binary with GOOS=windows
I have no idea if that makes any difference

**Reply by hajimehoshi** [reply]
Thanks. I tested cmd/chip8 after enabling vsync and using the default TPS:

![image](https://user-images.githubusercontent.com/16950/122648063-433cf280-d162-11eb-8955-33b1106c4aa9.png)

The FPS seems a little slow on Windows, but I didn't think this was significantly slow. TPS and FPS matched very well on macOS, so there might be some differences between Windows and POSIX implementation in Ebiten. I'll take a look further.

**Reply by hajimehoshi** [reply]
> The FPS seems a little slow on Windows, but I didn't think this was significantly slow. 

This was the same in Ebiten's examples, so this seems a usual thing.

In your code, there are some if-branches with detecting GOOS. What if removing these conditions?

**Reply by benc-uk** [reply]
> > The FPS seems a little slow on Windows, but I didn't think this was significantly slow.
> 
> This was the same in Ebiten's examples, so this seems a usual thing.
> 
> In your code, there are some if-branches with detecting GOOS. What if removing these conditions?

Those branches are very new (added just a few hours ago), just to handle cases with sound support

**Reply by hajimehoshi** [reply]
So, unfortunately I couldn't reproduce your problem on Windows.

* Did you execute WSL version and Windows version on the same machine? 
* Could you try rebooting your machine?

Thanks,

**Reply by benc-uk** [reply]
Yes it's all on the same machine

I think something more fundamental is happening
I added a crude counter to count how many loops my emulator processor loop is performing, and measure of the max it reaches over the last 2 seconds and display them under the FPS

On windows I get about 11 max cycles and on Linux I get 500+ playing the game CHIP-8 game INVADERS
I removed all the rendering code and I see the same problem, so I don't think it's the drawing rendering that is causing this, and possibly not Ebiten at all
![Screenshot 2021-06-19 173009](https://user-images.githubusercontent.com/14982936/122649344-5087bc00-d125-11eb-9998-bab159e4eae0.png)
![Screenshot 2021-06-19 173059](https://user-images.githubusercontent.com/14982936/122649357-62695f00-d125-11eb-80f9-c1d238d148eb.png)

**Reply by hajimehoshi** [reply]
Thanks! That's very interesting and I agree that might be a fundamental issue. Have you tried executing .exe via WSL, or Windows?

**Reply by benc-uk** [reply]
I run the Windows exe from a PowerShell terminal so it's running outside of WSL

**Reply by hajimehoshi** [reply]
Linux app shows 92 FPS, but what is your display's refresh rate?

**Reply by hajimehoshi** [reply]
My another guess is G-sync: https://github.com/hajimehoshi/ebiten/issues/1467

**Reply by benc-uk** [reply]
I think the culprit is the sleep I have inside my main emulator processor fetch/execute loop
If I remove it, the Windows version runs extremely fast (as you might expect)
If I add even a *very* small sleep e.g. `time.Sleep(time.Duration(1) * time.Nanosecond)` it slows to a crawl

I'm going to carry on investigating but for now I'm about 99% confident this is not an ebiten issue. Sorry for wasting your time.
If discover anything interesting or a solution I'll let you know

**Reply by hajimehoshi** [reply]
Aha. As far as I know, the precision of Windows timer is not very good and this can be more than 1-10 [ms] even though you specify 1[us] or less. It is surprising to me that WSL emulates more accurate timers for Linux.

> I'm going to carry on investigating but for now I'm about 99% confident this is not an ebiten issue. Sorry for wasting your time.

No problem, thank you for reporting!

**Reply by benc-uk** [reply]
I found this
https://github.com/golang/go/issues/44343

As for WSL, it's basically "full" Linux system now, just running inside a thin virtual machine, so to most purposes it works just like Linux

PS. Forgot to say, I love Ebiten, keep up the awesome work!

**Reply by benc-uk** [reply]
As a follow up, I've removed my separate processor loop and goroutine, and no longer rely on a sleep. I now do everything from within the ebiten Update loop, executing multiple CPU cycles inside that loop to speed things up. This has solved nearly all of the issues I was having with performance on Windows and even WASM is fully playable now

**Reply by hajimehoshi** [reply]
Great! I'm pretty happy to hear that your application with Ebiten worked well. Thank you for reporting!
